DOC: Added docs to clarify that MultiIndex.set_levels() interprets passed values as new components of the .levels attribute (#28294)

hweecat opened the pull request:
Added documentation on MultiIndex.set_levels to clarify that MultiIndex.set_levels() interprets passed values as new components of the .levels attribute

- [x] closes #28294 - or at least attempts to
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

Here's my first PR to pandas - looking forward to feedback on how I can further improve this.

hweecat committed:
DOC: added docs for MultiIndex.set_levels (#28294)

shortened lines to pass checks

hweecat committed:
DOC: added docs for MultiIndex.set_levels (#28294)

shortened lines to pass checks

jreback commented on doc/source/user_guide/advanced.rst:
```diff
@@ -177,6 +177,37 @@ completely analogous way to selecting a column in a regular DataFrame:
 See :ref:`Cross-section with hierarchical index <advanced.xs>` for how to select
 on a deeper level.
 
+.. _advanced.set_levels:
```
I would rather just edit the actual set_levels doc-string with examples. This is too much for the main docs (and not generally what someone would want).

hweecat committed:
edit set_levels docstring with additional examples

edit set_levels docstring with additional examples

edit set_levels docstring with additional examples

hweecat committed:
edit set_levels docstring with additional examples

edit set_levels docstring with additional examples

edit set_levels docstring with additional examples

jbrockmendel committed:
CLN: Exception*2 in groupby wrapper (#28771)

jbrockmendel committed:
TST: un-xfail 1 passing maybe_promote test (#28850)

tlynch1315 committed:
remove doc note about apply applying a function to the first element … (#28854)

jbrockmendel committed:
TST: un-xfail 22 tests (#28856)

rlamy committed:
TST: Improve compatibility with pypy error messages (#28844)

jbrockmendel committed:
TST: Fix not-boxed maybe_promote test (#28852)

jbrockmendel committed:
TST: fix 24 xfails in maybe_promote (#28833)

jbrockmendel committed:
CLN: Assorted cleanups (#28848)

WillAyd committed:
Clean up Abstract and Naming Definitions for GroupBy (#28847)

refactoriel committed:
DOC: Fix missing periods and non capitalized summary beginnings (#28858)

KimDoubleB committed:
fix unnecessary sort in pd.read_json and orient="index" (#28606)

jbrockmendel committed:
CLN: dont catch Exception in groupby var (#28883)

jbrockmendel committed:
CLN: dont catch on groupby.mean (#28878)

jbrockmendel committed:
CLN: assorted cleanups, remove unicode checks in cython (#28879)

jbrockmendel committed:
No catching needed for median (#28873)

kboone committed:
read_hdf closes HDF5 stores that it didn't open. (#28700)

shaido987 committed:
DOC: Fix commpiler typo in contributing.rst (#28891)

Corrects the spelling of compiler

jbrockmendel committed:
TST: Fix maybe_promote floating non-boxed tests (#28880)

jbrockmendel committed:
TST: Fix xfails for non-box maybe_promote on integer dtypes (#28864)

roshni-b committed:
pandas.io.formats.style.Styler.set_table_attributes: Parameter attributes type updated to str instead of string (#28887)

roshni-b committed:
DOC: Fixed PR06 errors in pandas.api.extensions.ExtensionArray (#28885)

jbrockmendel committed:
REF: simplify maybe_promote (#28899)

KimDoubleB committed:
Fix typo (#28915)

simonjayhawkins committed:
TYPING: lockdown test modules failing mypy (#28914)

jbrockmendel committed:
CLN: simplify maybe_promote in float and complex cases (#28913)

simonjayhawkins committed:
TYPING: errors reported by mypy 0.730 (#28910)

topper-123 committed:
clean tests/indexing/common.py (#28904)

jbrockmendel committed:
REF: use fused types for groupby_helper (#28886)

jjlkant committed:
BUG: Allow all int types for merge (GH28870) (#28875)

jbrockmendel committed:
TST: Fix 36 maybe_promote xfails wanting np.bytes_ instead of np.object_ (#28861)

simonjayhawkins committed:
TYPING: fix type annotation for pandas.io.formats.format._binify (#28908)

h-vetinari committed:
TST: add test coverage for maybe_promote (#23982)

jschendel committed:
TST: Use fixtures instead of setup_method for index tests (#28865)

jbrockmendel committed:
CLN: Exception (#28853)

jbrockmendel committed:
CLN: assorted cleanups, mostly post-black fixups (#28857)

tolaa001 committed:
DOC: PR06 docstring fixes (#28807)

ChiefMilesEdgeworth committed:
TST: Allow for multiple variables on the same line in docstring validation (#28811)

jbrockmendel committed:
CLN: try/except cleanups (#28939)

siddhesh committed:
BUG: Avoid undefined behaviour when converting from float to timedelta (#28918)

Summation of timedelta series with NaTs in them result in undefined
behaviour because the final wrapping step of the summation ends up
converting the NaNs in the sum through a direct cast to int64.  This
cast is undefined for NaN and just happens to work on x86_64 because
of the way cvttd2si works.  On Aarch64, the corresponding fcvtzs sets
the result to 0 on undefined input.

This fix trivially sets the conversion target to m8 instead of i8 so
that numpy correctly casts from NaN to NaT.  Note that the fix in
numpy for the same is pending in PR #numpy/numpy/14669 .

There is an existing test (test_sum_nanops_timedelta in
frame/test_analytics.py) that exercises this bug and has been verified
to have been fixed with this and the numpy patch.

S3rbane committed:
Added note to 'contributing.rst file', telling users to append GH Issue… (#28907)

jbrockmendel committed:
REF: maybe_promote refactor/cleanup (#28897)

topper-123 committed:
API: Add various deprecated attributes to ._deprecated (#28805)

ryankarlos committed:
TST: Test pivot_table() with categorical data (#28803)

timcera committed:
BUG: Need 'windows-1252' encoding for locale names. (#27368)

jorisvandenbossche committed:
DOC: disable nbsphinx including requirejs (#28940)

lukasbk committed:
CLN: fix mypy errors in pandas/tests/extension/test_numpy.py #28926 (#28947)

rohitsanj committed:
TST: add test_series_any_timedelta for GH17667 (#28942)

tolaa001 committed:
PR06 doc string fixes (#28946)

yogendrasoni committed:
fix #28926 pandas\api\test_api.py mypy errors (#28935)

*  fix #28926 pandas\api\test_api.py mypy errors

*  fix #28926 pandas\api\test_api.py mypy errors

*  changed to  type comment to support v 3.5

* removed section mypy-pandas.tests.api.test_api

* fix annotation and revert delete

* fix import sorting

* fix import sorting

dayoreke committed:
Eliminated _WriterBase class, removed unused fixtures from methods in pandas/io/excel/test_writers.py (#28753)

lrm25 committed:
Fix mypy errors for pandas\tests\* #28926 (test_algos.py) (#28960)

jb8429 committed:
DOC: Fixed PR08 and PR09 docstring errors in pandas.Series (#28845)

datapythonista commented on pandas/core/indexes/multi.py:
```diff
@@ -740,32 +748,51 @@ def set_levels(self, levels, level=None, inplace=False, verify_integrity=True):
+                                            (3, 'one'), (3, 'two')],
                                             names=['foo', 'bar'])
-        >>> idx.set_levels([['a', 'b'], [1, 2]])
+        >>> idx.set_levels([['a', 'b', 'c'], [1, 2]])
```
Why do you think an extra value makes this example clearer? I think a more concise example is easier to read.

hweecat replied:
My intention of adding an extra value at one of the levels (same for the _MultiIndex_ above) is to further highlight the differences in passing values to the two levels in the _MultiIndex_, for which the number of values in each level are not equal. Using different number of values in each level is to emphasize that _set_levels()_ passes values to indexes on 'by level' basis (by changing the _levels_ attribute), not on 'by row' basis.

Then again, if a more concise example could also explain sufficiently how _set_levels()_ work without ambiguity, I'm okay with reverting to the original _MultiIndex_.

datapythonista commented on pandas/core/indexes/multi.py:
```diff
@@ -740,32 +748,51 @@ def set_levels(self, levels, level=None, inplace=False, verify_integrity=True):
-        >>> idx.set_levels([['a', 'b'], [1, 2]], level=[0, 1])
+        >>> idx.set_levels([['a', 'b', 'c'], [1, 2]], level=[0, 1]).levels
+        FrozenList([['a', 'b', 'c'], [1, 2]])
+        >>> idx.set_levels([['a', 'b', 'c'], [1, 2, 3, 4]], level=[0, 1])
```
Can you separate this example from the previous, and explain what are we showing?

hweecat replied:
The intention of this example is to illustrate that _set_levels()_ passes values into the _levels_ attribute that is represented by a FrozenList containing list of values for each level in the _MultiIndex_, even when the number of values passed for a level is more than the number of indexes available in the _MultiIndex_ itself.

I'll separate this example from the previous example, and append the explanation to the example.

datapythonista reviewed (commented):
thanks @hweecat for working on this, added some comments.

ChiefMilesEdgeworth committed:
BUG: Partial fix for docstring validation for parameters (#28765)

HughKelley committed:
DOC: fix formatting in the ExtensionArray docstrings (#28686)

jorisvandenbossche committed:
BUG: use EA.astype in ExtensionBlock.to_native_types (#28841)

TomAugspurger committed:
Document 3.8 compatibility (#28982)

jbrockmendel committed:
CI: 3.8 build (#28730)

jbrockmendel committed:
remove unnecessary get_value_at calls (#28977)

jbrockmendel committed:
add types tot rank_1d_, rank_2d (#28978)

jbrockmendel committed:
CLN: Exception in _libs (#28967)

aaditya-panik committed:
CLN: pandas-dev#28926 Fix pandas/tests/tseries/offsets/test_offsets_properties (#28972)

aaditya-panik committed:
CLN: pandas-dev#28926 Fix mypy errors in pandas/tests/io/parser/conftest.py (#28973)

TomAugspurger committed:
DOC: 0.25.2 whatsnew cleanup (#29000)

andreasbuhr committed:
DOC: fix is_scalar documentation (#28998) (#29004)

TomAugspurger committed:
Setuptools CI fixup (#29010)

* CI: Fix miniconda upgrade issues

aaditya-panik committed:
CLN: Fix mypy errors in pandas/tests/io/test_sql.py Reverted and added type changes (#29006)

nrebena committed:
DOC: Reference level name as Term of HDFStore.select query (#28791) (#28793)

jbrockmendel committed:
CLN: move small bits outside of try/excepts (#28962)

WillAyd committed:
CLN: Consistent and Annotated Return Type of _iterate_slices (#28958)

jbrockmendel committed:
restore xfail (#29013)

Commit:
To html encoding add (#28692)

jbrockmendel committed:
add uint64 support for some libgroupby funcs (#28931)

batterseapower committed:
BUG: CategoricalIndex allowed reindexing duplicate sources (#28257)

topper-123 committed:
Separate MultiIndex names from levels (#27242)

oktaysabak committed:
DOC: fix code-block in the reshaping docs (#28838)

immaxchen committed:
CLN: Fix mypy error in pandas/tests/computation/test_eval.py (#29007)

topper-123 committed:
CLN: Clean DirNameMixin (#28957)

jbrockmendel committed:
REF: de-duplicate groupby_helper code (#28934)

saska committed:
CLN: fix mypy error pandas/tests/plotting/test_backend.py (#29016)

* CLN: fix mypy error pandas/tests/plotting/test_backend.py

GH28926, mypy didn't like setting non-existant attribute on module.
Dummy function, circumvented with __setattr__

jbrockmendel committed:
REF: re-raise AssertionError unchanged (#28959)

lukasbk committed:
tests/indexing/test_coercion.py typefix (#28990)

jbrockmendel committed:
DEPR: remove previously-deprecated broadcast/reduce kwargs from DataFrame.apply (#29017)

jbrockmendel committed:
BUG: Fix TypeError in _cython_agg_blocks (#29035)

jbrockmendel committed:
CI: xfail on numpy 1.18 (#29057)

* xfail on numpy 1.18

* CI: try using numpy wheel

jbrockmendel committed:
Stop catching TypeError in groupby methods (#29060)

TomAugspurger committed:
API: Restore getting name from MultiIndex level (#29061)

* API: Restore getting name from MultiIndex level

xref https://issues.apache.org/jira/browse/ARROW-6922 /
https://github.com/pandas-dev/pandas/pull/27242#issuecomment-543302582
/ https://github.com/pandas-dev/pandas/issues/29032

No docs yet, since it isn't clear how this will eventually sort out. But
we at least want to preserve this behavior for 1.0

* fixups

iamrajhans committed:
fixed issue of mypy for test_ujson (#29022)

jbrockmendel committed:
CLN: derivation of nogil param (#29047)

jbrockmendel committed:
REF: remove groupby_helper (#29040)

moi90 committed:
ENH: Informative dtype message for for assert_series_equal (#28993)

crepererum committed:
TST: add regression test for all-none-groupby (#29067)

Closes #21624

lrm25 committed:
Fix mypy errors for pandas\tests\*:  test_convert_to.py (#28965)

jbrockmendel committed:
CLN: catch less in groupby (#29077)

jbrockmendel committed:
CLN: simplify take_2d_multi (#29065)

jbrockmendel committed:
REF: use fused types in algos_rank_helper (#29044)

javad94 committed:
DOC: updated categorical docstring (#29068)

jbrockmendel committed:
CLN: tighten exception catching in indexes (#29078)

hweecat commented on pandas/core/indexes/multi.py:
```diff
@@ -719,8 +719,16 @@ def _set_levels(
-        new index.
+        Set levels on MultiIndex by passing a new value for each
+        index in the level. Defaults to returning new
+        index.
```
Oops, just realised I kinda broke PEP 257 for docstrings. Will make the changes.

hweecat commented on pandas/core/indexes/multi.py:
```diff
@@ -740,32 +748,51 @@ def set_levels(self, levels, level=None, inplace=False, verify_integrity=True):
-                                            (2, 'one'), (2, 'two')],
+                                            (2, 'one'), (2, 'two'),
+                                            (3, 'one'), (3, 'two')],
                                             names=['foo', 'bar'])
```
Sure, will apply this change.

hweecat committed:
explain set_levels().levels examples + formatting

hweecat committed:
explain set_levels().levels examples + formatting

resolve trailing whitespaces

hweecat committed:
explain set_levels().levels examples + formatting

resolve pep8 issues

explain set_levels().levels examples + formatting

hweecat committed:
explain set_levels().levels examples + formatting

resolve trailing whitespaces

hweecat committed:
explain set_levels().levels examples + formatting

pep8speaks commented:
Hello @hweecat! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/core/indexes/multi.py`](https://github.com/pandas-dev/pandas/blob/09fc1b4b86799a790449f0cd03f44e8b42f51973/pandas/core/indexes/multi.py):

> [Line 724:71](https://github.com/pandas-dev/pandas/blob/09fc1b4b86799a790449f0cd03f44e8b42f51973/pandas/core/indexes/multi.py#L724): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 725:79](https://github.com/pandas-dev/pandas/blob/09fc1b4b86799a790449f0cd03f44e8b42f51973/pandas/core/indexes/multi.py#L725): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 727:29](https://github.com/pandas-dev/pandas/blob/09fc1b4b86799a790449f0cd03f44e8b42f51973/pandas/core/indexes/multi.py#L727): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 728:1](https://github.com/pandas-dev/pandas/blob/09fc1b4b86799a790449f0cd03f44e8b42f51973/pandas/core/indexes/multi.py#L728): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 729:72](https://github.com/pandas-dev/pandas/blob/09fc1b4b86799a790449f0cd03f44e8b42f51973/pandas/core/indexes/multi.py#L729): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 730:71](https://github.com/pandas-dev/pandas/blob/09fc1b4b86799a790449f0cd03f44e8b42f51973/pandas/core/indexes/multi.py#L730): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 731:69](https://github.com/pandas-dev/pandas/blob/09fc1b4b86799a790449f0cd03f44e8b42f51973/pandas/core/indexes/multi.py#L731): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 732:66](https://github.com/pandas-dev/pandas/blob/09fc1b4b86799a790449f0cd03f44e8b42f51973/pandas/core/indexes/multi.py#L732): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 733:67](https://github.com/pandas-dev/pandas/blob/09fc1b4b86799a790449f0cd03f44e8b42f51973/pandas/core/indexes/multi.py#L733): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 790:73](https://github.com/pandas-dev/pandas/blob/09fc1b4b86799a790449f0cd03f44e8b42f51973/pandas/core/indexes/multi.py#L790): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 791:71](https://github.com/pandas-dev/pandas/blob/09fc1b4b86799a790449f0cd03f44e8b42f51973/pandas/core/indexes/multi.py#L791): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 792:71](https://github.com/pandas-dev/pandas/blob/09fc1b4b86799a790449f0cd03f44e8b42f51973/pandas/core/indexes/multi.py#L792): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 793:72](https://github.com/pandas-dev/pandas/blob/09fc1b4b86799a790449f0cd03f44e8b42f51973/pandas/core/indexes/multi.py#L793): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

SaturnFromTitan committed:
[#22550] Remove TestData from series-tests test_quantile.py (#29096)

hweecat committed:
Merge branches 'docfix-multiindex-setlevels' and 'docfix-multiindex-setlevels' of https://github.com/hweecat/pandas into docfix-multiindex-setlevels

explain set_levels().levels + formatting

jbrockmendel committed:
comments, catch less (#29088)

jbrockmendel committed:
REF: remove algos_rank_helper (#29086)

jbrockmendel committed:
CLN: Exception in DataFrame._reduce (#29085)

SaturnFromTitan committed:
[#22550] Remove TestData from series-tests test_operators.py (#29084)

jhereth committed:
TST: 2d index when constructing dataframe (#25416). (#29083)

jhereth committed:
TST: regression test for groupby with datetime and timedelta (#15562) (#29063)

jbrockmendel committed:
CLN: cython and docstring cleanups (#29089)

jbrockmendel committed:
BUG: parse_time_string failing to raise TypeError (#29098)

hweecat committed:
formatting changes

hweecat committed:
formatting changes

hweecat commented:
Travis CI is failing at Python 3.8-dev but passes for Python 3.6 and 3.7 - not too sure how to resolve this.

grigoriosgiann committed:
Add a regression test for the timezone issue (#29097)

AbhijeetKrishnan committed:
Fix typing errors (#29115)

Thanks, @AbhijeetKrishnan

AbhijeetKrishnan committed:
Fix typing errors (#29114)

hweecat committed:
Merge branch 'docfix-multiindex-setlevels' of https://github.com/hweecat/pandas into docfix-multiindex-setlevels

WillAyd commented:
@hweecat I'm not sure what happened but the git diff on this seems to have pulled in close to 200 files

Are you still fixing that up? It might be easier to start fresh with a new branch if unsure what happened

hweecat commented:
I think it could be because I tried to sync my fork with the upstream in order to pass Travis CI for Python 3.8 (Travis CI kept failing for Python 3.8 previously: https://travis-ci.org/pandas-dev/pandas/jobs/600207880).

Think I might need to start fresh with a new branch (from the master that is synced with the latest upstream) and add changes to that branch. Does that mean I might have to create a new pull request?

WillAyd commented:
Yea it might be easiest to just do that at this point

WillAyd commented:
Superseded by #29143
